external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+    fn handle_lvalue(&mut self, lval: &T::LValue) {
+        match &lval.value {
+            T::LValue_::Ignore => (),
+            T::LValue_::Var { ty, .. } => {
+                for (mident, name, loc) in qualified_datatype_name_of_type(ty) {
+                    if let Some(deprecation) = self.deprecated_type(mident, name) {
+                        self.register_deprecation_warning(deprecation.clone(), loc);
+                    }
+                }
+            }
+            T::LValue_::UnpackVariant(mident, dname, _, ty_params, fields)
+            | T::LValue_::BorrowUnpackVariant(_, mident, dname, _, ty_params, fields)
+            | T::LValue_::BorrowUnpack(_, mident, dname, ty_params, fields)
+            | T::LValue_::Unpack(mident, dname, ty_params, fields) => {
+                if let Some(deprecation) = self.deprecated_type(mident, dname) {
+                    self.register_deprecation_warning(deprecation.clone(), dname.loc());
+                }
+
+                for (mident, name, loc) in
+                    ty_params.iter().flat_map(qualified_datatype_name_of_type)
+                {
+                    if let Some(deprecation) = self.deprecated_type(mident, name) {
+                        self.register_deprecation_warning(deprecation.clone(), loc);
+                    }
+                }
+
+                for (_, _, (_, (ty, lval))) in fields.iter() {
+                    self.handle_lvalue(lval);
+                    for (mident, name, loc) in qualified_datatype_name_of_type(ty) {
+                        if let Some(deprecation) = self.deprecated_type(mident, name) {
+                            self.register_deprecation_warning(deprecation.clone(), loc);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Process all the deprecation warnings and add them to the environment.
+    fn process_and_add_deprecation_warnings(self) {
+        let attr_deprecation = |location| match location {
+            AttributePosition::Module => TypeSafety::DeprecatedModule,
+            AttributePosition::Constant => TypeSafety::DeprecatedConstant,
+            AttributePosition::Struct => TypeSafety::DeprecatedStruct,
+            AttributePosition::Enum => TypeSafety::DeprecatedEnum,
+            AttributePosition::Function => TypeSafety::DeprecatedFunction,
+            AttributePosition::AddressBlock
+            | AttributePosition::Use
+            | AttributePosition::Friend
+            | AttributePosition::Spec => unreachable!(
+                "ICE: Deprecations are not allowed in '{}'s and should have already been checked",
+                location
+            ),
+        };
+
+        for (DeprecationId { deprecation, .. }, locs) in self.deprecation_warnings {
+            let mut locs: Vec<_> = locs
+                .into_iter()
+                .flat_map(|(position, locs)| {
+                    Self::merge_locations(locs.into_iter().collect())
```

Comment:
Could you explain why this is necessary? bit lost

Reply:
I'll say, let's do whatever we can to make this _not_ necessary. Loc information should be roughly considered "private" by everything other than diagnostic rendering.

Reply:
Might not be necessary anymore -- I think this was papering over me picking the slightly incorrect location in one place. I believe it can be removed now.